File: tests/fixtures/auth/service.py
"""Authentication service."""

from typing import Optional

from .tokens import validate_token, generate_token, revoke_token
from .models import User


class BaseService:
    """Base service with common utilities."""

    def __init__(self):
        self._initialized = True

    def _log(self, message: str):
        print(f"[{self.__class__.__name__}] {message}")


class AuthService(BaseService):
    """Handles user authentication flows."""

    def __init__(self, db):
        super().__init__()
        self.db = db

    def login(self, email: str, password: str) -> Optional[str]:
        """Authenticate a user and return a token."""
        user = self._find_user(email)
        if user and user.check_password(password):
            self._log(f"Login successful for {email}")
            return generate_token(user)
        return None

    def logout(self, token: str) -> bool:
        """Revoke a user's token."""
        return revoke_token(token)

    def get_current_user(self, token: str) -> Optional[User]:
        """Get the user associated with a token."""
        return validate_token(token)

    def _find_user(self, email: str) -> Optional[User]:
        """Find a user by email."""
        return self.db.query(User).filter_by(email=email).first()


class AdminService(AuthService):
    """Extended auth service for admin operations."""

    def impersonate(self, admin_token: str, user_id: int) -> Optional[str]:
        """Allow admin to impersonate another user."""
        admin = self.get_current_user(admin_token)
        if admin and admin.is_admin:
            target = self.db.query(User).get(user_id)
            if target:
                return generate_token(target)
        raise PermissionError("Not authorized")

File: benchmarks/fixtures/webapp_php/auth/service.php
<?php

namespace App\Auth;

use App\Database\DatabaseConnection;
use App\Models\User;
use App\Utils\Logger;

use function App\Utils\get_logger;

/**
 * Base service with common utilities.
 */
class BaseService
{
    protected Logger $logger;

    public function __construct()
    {
        $this->logger = get_logger(static::class);
    }

    protected function log(string $message): void
    {
        $this->logger->info('[' . static::class . "] {$message}");
    }
}

/**
 * Handles user authentication flows.
 */
class AuthService extends BaseService
{
    public function __construct(private DatabaseConnection $db)
    {
        parent::__construct();
    }

    /**
     * Authenticate a user with email + password, returning a token on success.
     */
    public function login(string $email, string $password): ?string
    {
        $user = $this->findUser($email);
        if ($user !== null && $user->checkPassword($password)) {
            $this->log("Login successful for {$email}");

            return generate_token($user);
        }
        $this->log("Login failed for {$email}");

        return null;
    }

    public function logout(string $token): void
    {
        revoke_token($token);
    }

    public function getCurrentUser(string $token): User
    {
        return validate_token($token);
    }

    public function changePassword(string $token, string $oldPw, string $newPw): bool
    {
        $user = validate_token($token);
        if ($user->checkPassword($oldPw)) {
            $user->setPassword($newPw);

            return true;
        }

        return false;
    }

    private function findUser(string $email): ?User
    {
        return User::findByEmail($this->db, $email);
    }
}

/**
 * Extended auth service for admin operations.
 */
class AdminService extends AuthService
{
    public function impersonate(string $adminToken, int $userId): string
    {
        $admin = $this->getCurrentUser($adminToken);
        if ($admin->isAdmin) {
            $this->log("Admin {$admin->email} impersonating user {$userId}");

            return generate_token($admin);
        }
        throw new TokenError('Not authorized');
    }

    /**
     * @return list<User>
     */
    public function listAllUsers(string $adminToken): array
    {
        $admin = $this->getCurrentUser($adminToken);
        if ($admin->isAdmin) {
            return [];
        }
        throw new TokenError('Not authorized');
    }
}
